File: src/analyzer/classify/classify-ordinary-reference.ts
import { ReferenceFlags } from "../../ir/model.js";
import type { AstNode } from "../../ir/model.js";
import type { ClassifyResult } from "./classify-result.js";
import { isAstExpression } from "./is-ast-expression.js";
import { reference } from "./reference.js";

export function classifyOrdinaryReference(
  t: string,
  key: string | null,
  parent: AstNode,
): ClassifyResult {
  if (t === "AssignmentExpression" && key === "left") {
    const op = (parent as { operator?: string }).operator ?? "=";
    const flags =
      op === "="
        ? ReferenceFlags.Write
        : ReferenceFlags.Read | ReferenceFlags.Write;
    const right = parent["right"];
    return reference(flags, false, isAstExpression(right) ? right : null);
  }
  if (t === "UpdateExpression" && key === "argument") {
    return reference(ReferenceFlags.Read | ReferenceFlags.Write, false, null);
  }
  if (t === "CallExpression" && key === "callee") {
    return reference(ReferenceFlags.Read | ReferenceFlags.Call, false, null);
  }
  if (t === "NewExpression" && key === "callee") {
    return reference(ReferenceFlags.Read | ReferenceFlags.Call, false, null);
  }
  if (t === "MemberExpression" && key === "object") {
    return reference(
      ReferenceFlags.Read | ReferenceFlags.Receiver,
      false,
      null,
    );
  }
  let init = false;
  if (t === "VariableDeclarator" && key === "init") {
    init = true;
  }
  return reference(ReferenceFlags.Read, init, null);
}

File: crates/unsnarl-boundary-eslint-scope/src/classify/classify_ordinary_reference.rs
//! Classify an identifier reference that isn't in a binding /
//! skip slot.
//!
//! Matches on `AstKind::AssignmentExpression` and reads
//! `AssignmentExpression.operator: AssignmentOperator` directly to
//! pick the `WRITE` vs `READ_WRITE` flag combination.

use oxc_ast::AstKind;
use oxc_syntax::operator::AssignmentOperator;

use unsnarl_ir::reference::reference_flags::ReferenceFlags;

use crate::classify::classify_result::ClassifyResult;
use crate::classify::reference::reference;

pub(crate) fn classify_ordinary_reference(
    parent: &AstKind<'_>,
    key: Option<&'static str>,
) -> ClassifyResult {
    if let AstKind::AssignmentExpression(ae) = parent {
        if key == Some("left") {
            let flags = if matches!(ae.operator, AssignmentOperator::Assign) {
                ReferenceFlags::WRITE
            } else {
                ReferenceFlags::READ | ReferenceFlags::WRITE
            };
            return reference(flags, false);
        }
    }
    if matches!(parent, AstKind::UpdateExpression(_)) && key == Some("argument") {
        return reference(ReferenceFlags::READ | ReferenceFlags::WRITE, false);
    }
    let init = matches!(parent, AstKind::VariableDeclarator(_)) && key == Some("init");
    reference(ReferenceFlags::READ, init)
}

#[cfg(test)]
#[path = "classify_ordinary_reference_test.rs"]
mod classify_ordinary_reference_test;
